WalletWasabi/Crypto/Extensions.cs

```diff
@@ -1,11 +1,38 @@
 using System.Collections.Generic;
+using WalletWasabi.Helpers;
 using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Crypto.ZeroKnowledge.LinearRelation;
+using WalletWasabi.Crypto;
 
 namespace System.Linq
 {
 	public static class Extensions
 	{
 		public static GroupElement Sum(this IEnumerable<GroupElement> groupElements) =>
 			groupElements.Aggregate(GroupElement.Infinity, (ge, acc) => ge + acc);
+
+		public static IEnumerable<TResult> Zip<TFirst, TSecond, TThird, TResult>(this IEnumerable<TFirst> first, IEnumerable<TSecond> second, IEnumerable<TThird> third, Func<TFirst, TSecond, TThird, TResult> resultSelector)
+		{
+			Guard.NotNull(nameof(first), first);
+			Guard.NotNull(nameof(second), second);
+			Guard.NotNull(nameof(third), third);
+			Guard.NotNull(nameof(resultSelector), resultSelector);
+			using var e1 = first.GetEnumerator();
+			using var e2 = second.GetEnumerator();
+			using var e3 = third.GetEnumerator();
+			while (e1.MoveNext() && e2.MoveNext() && e3.MoveNext())
+			{
+				yield return resultSelector(e1.Current, e2.Current, e3.Current);
+			}
+		}
+
+		public static void CheckDimesions(this IEnumerable<Equation> equations, IEnumerable<ScalarVector> allResponses)
```

Comment:
Maybe it would be cleaner/more general to make this into `Guard.SameDimensions` which takes `IEnumerable<TFirst>` and `IEnumerable<TSecond>` and compares them. If `TFirst` and `TSecond` are `IEnumerable<>` themselves, it would recur, but otherwise they don't have to be the same type in the base case, just same `Count()`.

Using it would look something like:

```cs
Guard.SameDimensions(Equations.Select(x => x.Generators), allResponses));
```

Reply:
I find `Guard.SameDimensions` better than `CheckDimesions`.